lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -106,57 +115,199 @@ bool SessionStateEnforcer::m5g_init_session_credit(
-    SessionState& session_state) {
-  uint32_t l_teid = 0;
-  auto itp        = pdr_map_.equal_range(imsi);
+    std::unique_ptr<SessionState>& session_state,
+    SessionUpdate& session_update) {
+  uint32_t gnode_teid = 0;
+  bool gnb_teid_get   = true;
+  bool upf_teid_get   = false;
+
+  /* Check and update latest session rules
+   * we get gnodeb TEID, and IP address details here
+   */
+  auto session_id                        = session_state->get_session_id();
+  SessionStateUpdateCriteria& session_uc = session_update[imsi][session_id];
+  gnode_teid                             = update_session_rules(
+      imsi, session_state, session_uc, gnb_teid_get, upf_teid_get);
+  if (!gnode_teid) return false;
+  uint32_t cur_version = session_state->get_current_version();
+  session_state->set_fsm_state(CREATED, session_uc);
+  session_state->set_current_version(++cur_version, session_uc);
+  /* Reset the upf resend retransmission counter counter, send the session
+   * creation request to UPF
+   */
+  session_state->reset_rtx_counter();
+  m5g_send_session_request_to_upf(imsi, session_state);
+  return true;
+}
+
+/* Add defualt rules to the passed IMSI
+ */
+bool SessionStateEnforcer::add_default_rules(
+    std::unique_ptr<SessionState>& session, const std::string& imsi) {
+  // Check 2 default rules are added
+  if (pdr_map_.count(imsi) == DEFAULT_RULE_COUNT) return true;
+  /*
+   * Lets add the default uplink and downlink PDR rule to this
+   * imsi
+   */
+  if (!session->search_pdr(DEFAULT_DOWN_LINK_PDR_ID)) {
+    pdr_map_.insert(
+        std::pair<std::string, uint32_t>(imsi, DEFAULT_DOWN_LINK_PDR_ID));
+  }
+  if (!session->search_pdr(DEFAULT_UP_LINK_PDR_ID)) {
+    pdr_map_.insert(
+        std::pair<std::string, uint32_t>(imsi, DEFAULT_UP_LINK_PDR_ID));
+  }
+  /*
+   * Plus policy DB added anymore it would be part of
+   * this map
+   */
+  return true;
+}
+
+bool SessionStateEnforcer::insert_pdr_from_core(
+    const std::string& imsi, std::unique_ptr<SessionState>& session_state,
+    SessionStateUpdateCriteria& session_uc, SetGroupPDR& rule) {
+  uint32_t len, gnode_teid = 0;
+  char ip_str[INET_ADDRSTRLEN];
+  struct sockaddr_in ip_sa;
+  bool byte_zero_skip;
+  std::string ip_addr, teid;
+  // Get the latest session configuration
+  const auto& config = session_state->get_config();
+
+  teid.assign(config.rat_specific_context.m5gsm_session_context()
+                  .gnode_endpoint()
+                  .teid());
+  len            = teid.length();
+  gnode_teid     = 0;
+  byte_zero_skip = true;
+  for (uint32_t i = 0; i < len; i++) {
+    if (byte_zero_skip && (!teid[i])) continue;
+    // Rest continue;
+    byte_zero_skip = false;
+    gnode_teid <<= 8;
+    gnode_teid += teid[i];
+  }
+  // Set ip_str to NULL
+  *ip_str = '\0';
+  ip_addr.assign(config.rat_specific_context.m5gsm_session_context()
+                     .gnode_endpoint()
+                     .end_ipv4_addr());

```

Comment:
why would go through all these hoops? store the ip addresses as IP strings (e.g., 192.168.0.1) and teid as uint32_t in session context. Use the same convention on AMF, UPF and SMF. When mapping to AVPs or IEs then convert them into the format they desire...

Reply:
same as above

Reply:
Added the same, Raised the Issue #5679 -->#5679